### PR TITLE
Refactor cast scalars to use new function registry.

### PR DESCRIPTION
**Pull request by kovrus**
## Checklist

 - [ ] Added an entry in `CHANGES.txt` for user facing changes
 - [ ] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: sql/src/main/java/io/crate/expression/scalar/cast/CastFunction.java [hunk 1]
```diff
@@ -28,26 +28,119 @@
...
+        // We can drop the per type cast function already in 4.2. This change
+        // would require handling the metadata for the places where the old
+        // signature of the cast function were used, e.g. generated columns.
+        for (Map.Entry<String, DataType> function : CAST_SIGNATURES.entrySet()) {
+            module.register(
+                Signature.builder()
+                    .name(new FunctionName(null, function.getKey()))
+                    .kind(FunctionInfo.Type.SCALAR)
+                    .typeVariableConstraints(typeVariable("E"), typeVariable("V"))
+                    .argumentTypes(parseTypeSignature("E"), parseTypeSignature("V"))
```
Comment (kovrus) [inline sql/src/main/java/io/crate/expression/scalar/cast/CastFunction.java:62]: I have not changed the second argument, such is I restrict it `function.getValue().getTypeSignature()` then we won't much a function that takes as the second argument an object type with inner types. the generic `V` for the second argument works in this case. I also tried to drop coercion and explicitly provide the type info for the second argument, but then it loses the inner type info.

Comment (seut) [reply sql/src/main/java/io/crate/expression/scalar/cast/CastFunction.java:62]: We could  (and maybe must) add a special case for targetType = ObjectType with a more concrete signature of `parseTypeSignature("object(text, V)")` (including a variableArityGroup of `text, V`).
Currently all signature are the same and it will  always select the first one (with a probably wrong return type).
(This is btw. currently wrong, it should fail if multiple functions match, I'll provide a fix soon).
Loosing the innerType should not happen with your fix to `createType()` + `getTypeSignature()`.

File: common/src/main/java/io/crate/types/TypeSignature.java [hunk 2]
```diff
@@ -132,7 +132,13 @@ public String getBaseTypeName() {
             var builder = ObjectType.builder();
             for (int i = 0; i < parameters.size() - 1;) {
                 var valTypeSignature = parameters.get(i + 1);
-                builder.setInnerType(String.valueOf(i), valTypeSignature.createType());
+                final String innerTypeName;
+                if (valTypeSignature instanceof NamedTypeSignature) {
+                    innerTypeName = ((NamedTypeSignature) valTypeSignature).parameterName();
+                } else {
+                    innerTypeName = String.valueOf(i);
```
Comment (seut) [inline common/src/main/java/io/crate/types/TypeSignature.java:139]: This should never happen. We should add an assertion.

File: common/src/main/java/io/crate/types/TypeSignature.java [hunk 3]
```diff
@@ -160,7 +166,7 @@ public boolean equals(Object o) {
         if (this == o) {
             return true;
         }
-        if (o == null || getClass() != o.getClass()) {
```
Comment (seut) [inline common/src/main/java/io/crate/types/TypeSignature.java:163]: why is this removed?

File: common/src/main/java/io/crate/types/NamedTypeSignature.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.types;
+
+class NamedTypeSignature extends TypeSignature {
```
Comment (seut) [inline common/src/main/java/io/crate/types/NamedTypeSignature.java:25]: I would name this more concrete to avoid confusion. E.g. `ObjectParameterTypeSignature`.

File: common/src/main/java/io/crate/types/TypeSignature.java (hunk 3 above)
Comment (kovrus) [reply common/src/main/java/io/crate/types/TypeSignature.java:163]: I was about to ask a question here. Did it to be able to compare the parameters with NamedTypeSignature and TypeSignature, which should probably never be compared.

eg. parseSignature(object(text int)) is(ObjectType.getSignature) 

will adjust tests.

File: sql/src/test/java/io/crate/metadata/functions/SignatureBinderTest.java [hunk 5]
```diff
@@ -239,15 +241,19 @@ public void testMap() {
             .build();
 
         assertThat(getValueFunction)
-            .boundTo("object(text, bigint)", "text")
+            .boundTo(
+                ObjectType.builder()
+                    .setInnerType("V", DataTypes.LONG).build(), DataTypes.STRING)
```
Comment (seut) [inline sql/src/test/java/io/crate/metadata/functions/SignatureBinderTest.java:246]:
```suggestion
                    .setInnerType("V", DataTypes.LONG).build(),
                  DataTypes.STRING)
```

File: sql/src/test/java/io/crate/metadata/functions/SignatureBinderTest.java [hunk 6]
```diff
@@ -239,15 +241,19 @@ public void testMap() {
...
                 Map.of(
                     "K", type("text"),
                     "V", type("bigint"))
             ));
 
         assertThat(getValueFunction)
-            .boundTo("object(text, bigint)", "bigint")
+            .boundTo(
+                ObjectType.builder()
+                    .setInnerType("V", DataTypes.LONG).build(), DataTypes.LONG)
```
Comment (seut) [inline sql/src/test/java/io/crate/metadata/functions/SignatureBinderTest.java:256]:
```suggestion
                    .setInnerType("V", DataTypes.LONG).build(),
                  DataTypes.LONG)
```

File: sql/src/main/java/io/crate/expression/scalar/cast/CastFunction.java (hunk 1 above)
Comment (kovrus) [reply sql/src/main/java/io/crate/expression/scalar/cast/CastFunction.java:62]: it is not possible to parse `parseTypeSignature("object(text, V)")` and call create type on it later on due to the assertion in the `createType` method. Also, there are probably also other cases, like `array(object)`, etc.

Hm but all function signatures are different, coz they encode the return type name into their names. For object type matching `parseTypeSignature("V")` works fine, and inner types are still there.

With this signature matching for two arguments here, it feels that we are trying to provide a correct solution on top of the existing workaround for the return type in the CastFunctionResolver.

Comment (seut) [reply sql/src/main/java/io/crate/expression/scalar/cast/CastFunction.java:62]: > it is not possible to parse parseTypeSignature("object(text, V)") and call create type on it later on due to the assertion in the createType method

Right but this should (and currently will) never happen. `createType()` should only be called on bound argument types which won't be the defined type signatures.

> Hm but all function signatures are different, coz they encode the return type name into their names.

Ah right, missed the different function names.
As we've discussed this looks like we should get rid of these different function names eventually as all will end up in the same signature + implementation. But yes, lets do this later.

File: sql/src/main/java/io/crate/expression/scalar/cast/CastFunction.java [hunk 7]
```diff
@@ -28,26 +28,119 @@
...
+                    .returnType(function.getValue().getTypeSignature())
+                    .forbidCoercion()
+                    .build(),
+                args -> {
+                    DataType<?> sourceType = args.get(0);
+                    DataType<?> targetType = args.get(1);
+                    if (!sourceType.isConvertableTo(targetType)) {
+                        throw new ConversionException(sourceType, targetType);
+                    }
```
Comment (seut) [inline sql/src/main/java/io/crate/expression/scalar/cast/CastFunction.java:71]: this check can be removed, the signature binder will already take care of this (if not its a bug ;-).

File: sql/src/main/java/io/crate/expression/scalar/cast/CastFunction.java [hunk 8]
```diff
@@ -28,26 +28,119 @@
...
+                    .argumentTypes(parseTypeSignature("E"))
+                    .returnType(function.getValue().getTypeSignature())
+                    .forbidCoercion()
+                    .build(),
+                args -> {
+                    DataType<?> sourceType = args.get(0);
+                    DataType<?> targetType = function.getValue();
+                    if (!sourceType.isConvertableTo(targetType)) {
+                        throw new ConversionException(sourceType, targetType);
+                    }
```
Comment (seut) [inline sql/src/main/java/io/crate/expression/scalar/cast/CastFunction.java:98]: can be removed, see previous comment

File: sql/src/main/java/io/crate/expression/scalar/cast/CastFunction.java (hunk 7 above)
Comment (kovrus) [reply sql/src/main/java/io/crate/expression/scalar/cast/CastFunction.java:71]: hm looked at it a bit, I am not sure whether it is in the scope of the signature binding. It looks like it is part of the cast logic, checking whether the first argument of the function can be converted to the second. Maybe it can be even moved into the CastFunctionReolver and checked based in tryCast flag.

Comment (seut) [reply sql/src/main/java/io/crate/expression/scalar/cast/CastFunction.java:71]: Ah right, these are 2 different type variables.
